### PR TITLE
feat: add support for interests in the profile screen + tests

### LINKED ISSUE #112: Add support for interests in profile
Be able to add some interests on the profile page.
Image

### PR DESCRIPTION
**Pull request by BogdanCDev**
## Summary

This PR adds support for managing user interests within the app, allowing users to view, add, and remove their interests. Interests are now displayed both in the ProfileScreen and EditProfileScreen. The implementation includes UI components for displaying and editing interests, as well as unit and UI tests to ensure functionality and reliability. Significant enhancements have also been made to the UserViewModel and related classes to support these new features. The associated tests have been updated and expanded to cover the new functionalities.

## Changes

- **Screens/UI:** 
  - ProfileScreen:
    - Interest Display:
      - Updated to display the user's interests as chips, providing a personalized overview of the user's profile.
      - Used the InterestChip composable to render each interest.
  - EditProfileScreen:
    - Interest Management:
      - Implemented UI components for displaying the user's current interests as editable chips.
      - Added functionality to add new interests through an input field with validation to prevent duplicates.
      - Included the ability to remove interests via a delete icon on each interest chip, with a confirmation dialog to prevent accidental deletions.
    - Keyboard Interaction:
      - Enhanced keyboard interactions, including dismissing the keyboard when "Done" is pressed in "Add interest" text field.
    - Permission Handling:
      - Updated UI to handle permissions when choosing a profile picture, including requesting runtime permissions and showing rationale dialogs.
- InterestChip and InterestChipEditable:
  - Created new composables to represent interests:
    - InterestChip displays interests in the ProfileScreen.
    - InterestChipEditable is used in EditProfileScreen with additional functionality for editing.

- **Bug Fixes/Improvements:**
  - This PR solves #112 

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [X] This PR resolves #112 
- [X] Tests have been added for new functionality.
- [X] Screenshots or UI changes have been attached.
- [X] Documentation has been updated.

##  Testing

- Manual Testing:
  - Tested on emulator and physical devices.
  - Verified that interests are displayed correctly in the ProfileScreen.
  - Verified adding and removing interests functions correctly in EditProfileScreen.
  - Tested permission requests and rationale dialogs when changing the profile picture.

- UI Tests:
  - Implemented UI tests in ProfileScreenTest to verify interests display.
  - Implemented UI tests in EditProfileScreenTest for adding and removing interests.
  - Tested the confirmation dialog for interest removal.
  - Verified the behavior of the keyboard interactions.

- Unit Tests:
  - Added unit tests for InterestChip and InterestChipEditable composables.
  - Ensured that the UI components display correctly and handle user interactions as expected.

##  Related Issues/Tickets

This PR closes #112 

## Screenshots (if applicable)

Screenshot 2024-11-07 at 22 49 02
 
Screenshot 2024-11-07 at 22 49 08

Screenshot 2024-11-07 at 22 49 15

Screenshot 2024-11-07 at 22 49 32

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/ui/profile/EditProfileScreen.kt [hunk 1]
```diff
@@ -160,3 +317,14 @@ fun EditProfileScreen(userViewModel: UserViewModel, navigationActions: Navigatio
...
       })
 }
+
+fun Context.findActivity(): ComponentActivity {
+  var context = this
+  while (context is ContextWrapper) {
+    if (context is ComponentActivity) {
+      return context
+    }
+    context = (context as ContextWrapper).baseContext
```
Comment (1andreea26) [inline app/src/main/java/com/android/voyageur/ui/profile/EditProfileScreen.kt:327]: This here is marked for me as useless cast, maybe you could remove it.

File: app/src/androidTest/java/com/android/voyageur/ui/profile/ProfileScreenTest.kt [hunk 2]
```diff
@@ -160,4 +169,31 @@ class ProfileScreenTest {
     // Assert: Verify that the navigation was triggered
     verify(navigationActions).navigateTo(Route.AUTH)
   }
+
```
Comment (1andreea26) [inline app/src/androidTest/java/com/android/voyageur/ui/profile/ProfileScreenTest.kt:172]: You could also add a test which has null or white spaces as interest and check that it doesn't get saved (I checked manually and it works).

**Review by 1andreea26 — CHANGES_REQUESTED** [review]
Great job on the interests! They work really well and the tests are well-made. However permissions for the profile picture don't work  correctly (if the user selects "Don't allow access" from the settings of the phone, it still opens the gallery in the app and allows you to upload a photo), so I would not merge them into main at this time. Moreover, the PR is a bit long, it could be great if next time you split it a bit more :)

**Review by eleonoraka — CHANGES_REQUESTED** [review]
Thanks for implementing the user interests feature! The implementation looks solid, with good attention to UI components, error handling, and user experience. 
However, I noticed that the code coverage for EditProfileScreen.kt is currently below our 80% minimum threshold. While you've added some great tests already, we'll need a bit more coverage before merging.
Once we get the coverage up to 80% and solve the issues brought up, this will be ready to merge. Great work on this feature overall!

**Comment by BogdanCDev** [comment]
> Great job on the interests! They work really well and the tests are well-made. However permissions for the profile picture don't work correctly (if the user selects "Don't allow access" from the settings of the phone, it still opens the gallery in the app and allows you to upload a photo), so I would not merge them into main at this time. Moreover, the PR is a bit long, it could be great if next time you split it a bit more :)

Thank you for your review. Regarding the permissions issue with the profile picture, I haven't been able to reproduce this problem on my device. It might be related to emulator settings or residual permissions from previous tests. I'll attach a screen recording to demonstrate what I'm seeing on my end. 

https://github.com/user-attachments/assets/16c5f0a5-853a-4918-ac89-d43e75efcece

File: app/src/main/java/com/android/voyageur/ui/profile/EditProfileScreen.kt (hunk 1 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/profile/EditProfileScreen.kt:327]: You are right! Thank you!

File: app/src/androidTest/java/com/android/voyageur/ui/profile/ProfileScreenTest.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/androidTest/java/com/android/voyageur/ui/profile/ProfileScreenTest.kt:172]: Thank you very much for the suggestion!

Comment (1andreea26) [reply app/src/androidTest/java/com/android/voyageur/ui/profile/ProfileScreenTest.kt:172]: I noticed you added the test, thanks!

**Comment by 1andreea26** [comment]
> > Great job on the interests! They work really well and the tests are well-made. However permissions for the profile picture don't work correctly (if the user selects "Don't allow access" from the settings of the phone, it still opens the gallery in the app and allows you to upload a photo), so I would not merge them into main at this time. Moreover, the PR is a bit long, it could be great if next time you split it a bit more :)
> 
> Thank you for your review. Regarding the permissions issue with the profile picture, I haven't been able to reproduce this problem on my device. It might be related to emulator settings or residual permissions from previous tests. I'll attach a screen recording to demonstrate what I'm seeing on my end.
> 
>  Screen.Recording.2024-11-09.at.17.37.48.mov

There might have been an issue on my side. Now everything works as it should! Thanks for adding the video.

**Comment by BogdanCDev** [comment]
> > > Great job on the interests! They work really well and the tests are well-made. However permissions for the profile picture don't work correctly (if the user selects "Don't allow access" from the settings of the phone, it still opens the gallery in the app and allows you to upload a photo), so I would not merge them into main at this time. Moreover, the PR is a bit long, it could be great if next time you split it a bit more :)
> > 
> > 
> > Thank you for your review. Regarding the permissions issue with the profile picture, I haven't been able to reproduce this problem on my device. It might be related to emulator settings or residual permissions from previous tests. I'll attach a screen recording to demonstrate what I'm seeing on my end.
> > Screen.Recording.2024-11-09.at.17.37.48.mov
> 
> There might have been an issue on my side. Now everything works as it should! Thanks for adding the video.

Thank you very much for your review!

**Review by 1andreea26 — APPROVED** [review]
Behaviour of the Profile Screen and Edit Profile Screen work as expected and tests have been added to help with the line coverage. I will approve this. Good job!
